src/sage/graphs/graph_decompositions/modular_decomposition.pyx

```diff
@@ -1,27 +1,161 @@
+        sage: recreate_decomposition(15, corneil_habib_paul_tedder_algorithm,
+        ....:                                                         3, 4, 0.2)
+        sage: recreate_decomposition(10, corneil_habib_paul_tedder_algorithm,
+        ....:                                                         4, 5, 0.2)
+        sage: recreate_decomposition(3, corneil_habib_paul_tedder_algorithm,
+        ....:                                                         6, 5, 0.2)
+    """
+    cdef CGraphBackend Gbackend = <CGraphBackend> G._backend
+    cdef CGraph cg = Gbackend.cg()
+
+    cdef vector[int] sigma
+    cdef vector[vector[int]] lex_label
+    cdef vector[size_t] xslice_len
+
+    # Compute the slice decomposition using the extended lexBFS algorithm
+    extended_lex_BFS(cg, sigma, NULL, -1, NULL, &xslice_len, &lex_label)
+
+    cdef SDData SD
+    SD.set_from_data(0, sigma.data(), xslice_len.data(), lex_label.data())
+
+    MD = corneil_habib_paul_tedder_inner(SD)
+
+    r = md_tree_node_to_md_tree(MD, Gbackend)
+    dealloc_md_tree_nodes_recursively(MD)
+    return r
+
+
+cdef object _md_tree_node_to_md_tree_inner_rec(const md_tree_node *n,
+                                               CGraphBackend Gb):
+    """
+    Utility function for :func:`md_tree_node_to_md_tree`.
+    """
+    cdef md_tree_node *c
+    if deref(n).is_leaf():
+        return Node.create_leaf(Gb.vertex_label(deref(n).vertex))
+    else:
+        if deref(n).is_series():
+            node = Node(NodeType.SERIES)
+        elif deref(n).is_parallel():
+            node = Node(NodeType.PARALLEL)
+        else:  # is_prime
+            node = Node(NodeType.PRIME)
+        node.children.extend(
+            _md_tree_node_to_md_tree_inner_rec(c, Gb)
+                                                  for c in deref(n).children)
+        return node
+
+
+cdef object md_tree_node_to_md_tree(const md_tree_node *n, CGraphBackend Gb):
+    """
+    This function converts a modular decomposition tree (given as a pointer to a
+    md_tree_node) into an object of the Python class Node (which is then
+    converted into the correct output by :meth:`Graph.modular_decomposition`).
+
+    The graph backend is needed to convert the int stored into the md_tree_node
+    into the corresponding vertex of the Python graph.
+
+    This function deals with the case of an empty tree and then delegates the
+    actual conversion to :func:`_md_tree_node_to_md_tree_inner_rec`.
+
+    TESTS:
+
+    Indirect doctests::
+
+        sage: from sage.graphs.graph_decompositions.modular_decomposition import *
+        sage: corneil_habib_paul_tedder_algorithm(Graph(1))
+        NORMAL [0]
+        sage: corneil_habib_paul_tedder_algorithm(Graph(2))
+        PARALLEL [NORMAL [0], NORMAL [1]]
+        sage: corneil_habib_paul_tedder_algorithm(graphs.CompleteGraph(3))
+        SERIES [NORMAL [1], NORMAL [2], NORMAL [0]]
+    """
+    if n == NULL:
+        return Node(NodeType.EMPTY)
+    else:
```

Comment:
remove `else:`

Reply:
done in [451b325](https://github.com/sagemath/sage/pull/39038/commits/451b3256ac2cbbb00d2ff24ecc80339fe73d5e40)